feat: add new package for downsampling data

Rahul-Vasan opened the pull request:
Added a new package containing functions for polyline simplification. The RDP algorithm enables the downsampling of data by retaining the variations in the curve to create visually indistinguishable graphs.

### Checklist

Dear Author :wave:, the following checks should be completed (or explicitly dismissed) before merging.

- [x] ✏️ Write a PR description, regardless of triviality, to include the _value_ of this PR
- [x] 🔗 Reference related issues
- [x] 🏃 Test cases are included to exercise the new code
- [x] 🧪 If **new packages** are being introduced to stdlib, link to Working Group discussion notes and ensure it lands under `experimental/`
- [x] 📖 If **language features** are changing, ensure `docs/Spec.md` has been updated

Dear Reviewer(s) :wave:, you are responsible (among others) for ensuring the completeness and quality of the above before approval.

skartikey commented:
@Rahul-Vasan This looks good. I have already reviewed this [here](https://github.com/influxdata/flux/pull/5027) Is there anything changed since then?

Rahul-Vasan commented:
> @Rahul-Vasan This looks good. I have already reviewed this [here](https://github.com/influxdata/flux/pull/5027) Is there anything changed since then?

Rahul-Vasan commented:
> @Rahul-Vasan This looks good. I have already reviewed this [here](https://github.com/influxdata/flux/pull/5027) Is there anything changed since then?

Nothing has changed Sunil. I have just addressed the last unaddressed comments from the previous PR

sanderson commented on stdlib/polylinesimp/polylinesimp.flux:
```diff
@@ -0,0 +1,65 @@
+// rdp (Ramer Douglas Peucker algorithm) is a polyline simplification function that decimates a curve composed of line segments into a downsampled yet visually indistinguishable curve with fewer points.
+//
+// ## Parameters
+// - valColumn: column that corresponds to the Y axis of the given curve. (optional) (default valColumn: _value)
+// - timeColumn: column that corresponds to the X axis of the given curve. (optional) (default timeColumn: _time)
+// - epsilon: user defined maximum tolerance value that determines the amount of compression. (optional) (epsilon should be greater than 0.0)
+// - retention: user defined retention rate, which specifies the percentage of points to be retained after downsampling. (optional) (Retention rate should be between 0.0 and 100.0)
+// - tables: Input data. Default is piped-forward data (`<-`).
```
```suggestion
// - valColumn: Column with Y axis values of the given curve. Default is `_value`.
// - timeColumn: Column with X axis values of the given curve. Default is `_time`.
// - epsilon: Maximum tolerance value that determines the amount of compression. Default is ??.
//
//   Epsilon should be greater than or equal to `0.0`.
//
// - retention: Percentage of points to retain after downsampling. Default is ??.
//
//   Retention rate should be between `0.0` and `100.0`.
//
// - tables: Input data. Default is piped-forward data (`<-`).
```

- What is the default epsilon?
- What is the default retention rate?

Rahul-Vasan replied:
Both these parameters dont have default values, so if both of them are not provided we run the algorithm in a mode where is automatically learns a good estimate of epsilon for the given data.

sanderson commented on stdlib/polylinesimp/polylinesimp.flux:
````diff
@@ -0,0 +1,65 @@
+//
+// ### Downsample the data from abcd.csv using the epsilon value 1.5
+// ```
+// # import "polylinesimp"
+// # import "internal/gen"
+// #
+// #  gen.tables(n: 16, seed: 1234)
+// #          |> polylinesimp.rdp(valColumn: "_value",timeColumn: "_time", epsilon:1.5)
+// #
````
```suggestion
// # import "internal/gen"
// import "polylinesimp"
//
// # data = gen.tables(n: 16, seed: 1234)
// #
// < data
// >     |> polylinesimp.rdp(epsilon: 1.5)
```

sanderson commented on stdlib/polylinesimp/polylinesimp.flux:
```diff
@@ -0,0 +1,65 @@
+//
+// ## Examples
+//
+// ### Downsample the data from abcd.csv using the epsilon value 1.5
```
Where is `abcd.csv` coming from?

```suggestion
// ### Downsample the data using a 1.5 epsilon value
```

sanderson commented on stdlib/polylinesimp/polylinesimp.flux:
````diff
@@ -0,0 +1,65 @@
+// #
+// ```
+//
+// ### Downsample the data from abcd.csv using the retention rate of 90%
+// ```
+// # import "polylinesimp"
+// # import "internal/gen"
+// #
+// #  gen.tables(n: 16, seed: 1234)
+// #          |> polylinesimp.rdp(valColumn: "_value",timeColumn: "_time", retention:90.0)
+// #
+// ```
````
```suggestion
// ### Downsample the data using a 90% retention rate
// ```
// # import "internal/gen"
// import "polylinesimp"
//
// #  data = gen.tables(n: 16, seed: 1234)
// #
// < data
// >     |> polylinesimp.rdp(retention: 90.0)
// ```
```

sanderson commented on stdlib/polylinesimp/polylinesimp.flux:
````diff
@@ -0,0 +1,65 @@
+// ```
+//
+// ## Metadata
+// introduced: 0.7.0
````
This metadata value will be inherited from the package.

```suggestion
```

onelson commented on libflux/go/libflux/buildinfo.gen.go:
```diff
@@ -325,6 +325,8 @@ var sourceHashes = map[string]string{
 	"stdlib/planner/window_push_test.flux":                                                        "779f83caca79fec5e4129cb6d1b05b2fbcbb45b479acab69e075ad2ad62d9df2",
 	"stdlib/planner/window_sum_eval_test.flux":                                                    "460d45de75419da033d9bdf0b9a8c0d1bbdb57ed0d02f44d5f55d4d4b45086f1",
 	"stdlib/planner/window_sum_push_test.flux":                                                    "eb8a9bb757503a40da2f397acce7fda67cd71419f1eae420fb3c171d2f01d185",
+	"stdlib/polylinesimp/polylinesimp.flux":                                                       "14e941d1205a2ea89f4e01ea5c38ad4d70b35966745741397c4d02668d0ee279",
+	"stdlib/polylinesimp/polylinesimp_test.flux":                                                  "e1437db6cda53862420571117c90a346f45118feb82c202106a5204cde51e97d",
```
Per the checklist:  
> If new packages are being introduced to stdlib, link to Working Group discussion notes and ensure it lands under experimental/

Should this go into `experimental/` to start in case the team wants to fuss with the public API?

Rahul-Vasan replied:
Thanks @onelson will move it to ### experimental/

nathanielc commented on stdlib/polylinesimp/polylinesimp.flux:
```diff
@@ -0,0 +1,65 @@
+// ## Metadata
+// introduced: 0.7.0
+//
+package polylinesimp
```
Should we call this just `polyline` ? That still communicates what it does but is a bit shorter?

Additionally this should be in the `experimental` parent package, that gives us a chance to update and iterate on the API a bit.

Rahul-Vasan replied:
Sure!

nathanielc commented on stdlib/polylinesimp/polylinesimp.flux:
```diff
@@ -0,0 +1,65 @@
+// - valColumn: column that corresponds to the Y axis of the given curve. (optional) (default valColumn: _value)
+// - timeColumn: column that corresponds to the X axis of the given curve. (optional) (default timeColumn: _time)
+// - epsilon: user defined maximum tolerance value that determines the amount of compression. (optional) (epsilon should be greater than 0.0)
+// - retention: user defined retention rate, which specifies the percentage of points to be retained after downsampling. (optional) (Retention rate should be between 0.0 and 100.0)
```
What happens if both epsilon and retention are specified? 

What is the default if neither is specified?

Rahul-Vasan replied:
@nathanielc If both epsilon and retention are provided we exit with an error message **"Please provide either epsilon or retention rate and not both"**

if both epsilon and retention are not specified the algorithm will run the automatic case and try to make an educated guess for the epsilon value.

nathanielc commented on stdlib/polylinesimp/polylinesimp.flux:
```diff
@@ -0,0 +1,65 @@
+// Package polylinesimp provides methods for polyline simplication which is an efficient way of downsampling curves retaining the moments of variation throughout the path.
+// This class of algorithms enable efficient rendering of graphs and visualizations without having to load the entire corpus of data points into memory.
+// This is done by reducing the number of vertices that donot contribute significantly to the convexity and concavity of the shape.
```
`donot` -> `do not`

nathanielc commented on stdlib/polylinesimp/polylinesimp.flux:
````diff
@@ -0,0 +1,65 @@
+// #
+// ```
+//
+// ### Downsample the data from abcd.csv by automatically calculating the maximum tolerance beyond which producing a visually indistingushable curve will not be possible. This can be used when both epsilon and the retention rate are tricky to be judged.
````
`abcd.csv` doesn't exist anymore, should this be updated?

nathanielc commented on stdlib/polylinesimp/rdp/rdp.go:
```diff
@@ -0,0 +1,227 @@
+func (r *Rdp) Do(vs *array.Float, ts *array.Float) (*array.Int, *array.Float, error) {
+	r.vs = vs
+	r.ts = ts
+	points := r.ToPoints()
```
This logic invalidates the gains had by using the arrow array values. Is it possible to write the algorithm in terms of `r.vs` and `r.ts` using a shared index? That way we do not have to allocate the data multiple times?

nathanielc reviewed (changes requested):
Some comments below. Overall looks great!
